### ISSUE TITLE
Template inference script broken

### ISSUE THREAD

**Issue by pplantinga** [speechbrain/speechbrain#2701]
### Describe the bug

When trying to run inference on `templates/speech_recognition/ASR` I get the following error:

```
FileNotFoundError: [Errno 2] No such file or directory: '/home/pplantinga/Documents/Repositories/speechbrain/docs/tutorials/nn/speechbrain/templates/speech_recognition/ASR/lm.ckpt'
```

The inference class is looking for the model file `lm.ckpt` in the `source` folder rather than the `savedir` folder, which is different behavior from before. Perhaps this is related to the changes to `fetch` in #2476 but I haven't looked into this enough to be sure. Perhaps this is worth looking into more so we can at the very least alert users that this behavior is different, and possibly consider changes to revert to old behavior.

### Expected behaviour

Running the template inference file should not crash.

### To Reproduce

```
cd templates/speech_recognition/ASR/
python train.py train.yaml
python transcribe_file.py audio.wav results/...
```

**Comment by pplantinga** [comment]
Follow-up issue: the LM and tokenizer are no longer linked by default, meaning that even with this issue fixed, we will still have another issue which is to make these linked in the save folder again so we can find them later

**Comment by pplantinga** [comment]
Which brings up the issue, should we keep the new default (NO_LINK) for pretrainer or should we revert this to SYMLINK

https://speechbrain.readthedocs.io/en/latest/API/speechbrain.utils.parameter_transfer.html#speechbrain.utils.parameter_transfer.Pretrainer.collect_files

**Comment by asumagic** [comment]
> When I suggested changing the default to NO_LINK I was imagining that the files would be downloaded directly to the target directory rather than to a cache directory, not sure how easy/possible it is to do this with huggingface.

The option closest to the semantics you describe is `COPY_SKIP_CACHE`:

```
If the file is remote and not in cache, fetch it, preferably directly to the destination directory.
Then, create a copy in the destination folder to the local file, if necessary.
```

Defaulting to copies from cache is probably not great, though.

I wonder if the following semantics **compared to v1.0.1** (not the current *develop*) would be overall better:

- Code that explicitly sets the `fetch` `savedir` behave the same (default to `SYMLINK`).
- Code can now explicitly opt in to any of the new strategy options.
- Code that does not set `savedir` are now assumed to not rely on where the files are saved (default and enforce `NO_LINK`).
    - Local paths returned as is, no symlink
    - Unsure about URLs:
        - Error out? (More logical but more manual intervention required)
        - Keep the previous default? (Less consistent)
        - Introduce a new centralized `~/.cache/speechbrain`? Would be better if we could avoid it though IMO
    - HF returns from HF cache, no symlink
- Examples and uses in the codebase edited to not specify `savedir`.
- Pretrained/etc. would also default to no `savedir` instead of the current hashed path?
- Ensure that at least one message sums up what any fetch operation has resulted into, at INFO level.

One annoyance with this approach is that existing setups get left with a misleading cache directory that would no longer be used.

**Comment by pplantinga** [comment]
Some of these things should probably be implemented when we have the chance:

Some of these things should probably be implemented when we have the chance:

Some of these should definitely get implemented when we get a chance:

> I wonder if the following semantics **compared to v1.0.1** (not the current _develop_) would be overall better:
> 
> * Code that explicitly sets the `fetch` `savedir` behave the same (default to `SYMLINK` or perhaps an environment variable).

Agree here, this is what's weird about the templates -- they set savedir but then it is empty.

> * Code can now explicitly opt in to any of the new strategy options.

Isn't this already possible? Or are you suggesting we add an option to Pretrainer init so that this can be done in YAML in addition to the possibility to specify in the `.collect_files()` call? I was thinking about the same.

> * Code that does not set `savedir` are now assumed to not rely on where the files are saved (default and enforce `NO_LINK`).

I also feel this is what I would expect but I'm worried about the complexity / transparency of having one parameter (the link strategy) depend on another (the savedir). I suppose the simplest way would be to just say link strategy has no effect unless savedir is also specified.

> Unsure about URLs:

I don't have much opinion about URLs, I'd be fine with keeping the default as is.

> One annoyance with this approach is that existing setups get left with a misleading cache directory that would no longer be used.

I guess this would apply if `savedir` was not specified and the cache was the default cache directory. There could be a weird error if they depend on those files being up-to-date and we updated a recipe and the updates aren't reflected in the stale cache, but honestly I would expect that to be rare. If we're concerned about it, we could add a function to check if the cache dir exists and emit a warning.

----

Overall I'd be happy with this approach, should I go ahead and add it to the template fix PR? Or no, it would probably be better as a separate PR yeah?

**Comment by asumagic** [comment]
> Isn't this already possible?

Yes, the list was intended to be semantics vs v1.0.1, not vs the (merged) symlink PR.

> Or are you suggesting we add an option to Pretrainer init so that this can be done in YAML in addition to the possibility to specify in the .collect_files() call? I was thinking about the same.

No but that would be good :)

> I also feel this is what I would expect but I'm worried about the complexity / transparency of having one parameter (the link strategy) depend on another (the savedir). I suppose the simplest way would be to just say link strategy has no effect unless savedir is also specified.

Unsure. This doesn't really change what combination of parameters is ultimately meaningful.

> I don't have much opinion about URLs, I'd be fine with keeping the default as is.

Ok, might make sense.

> I guess this would apply if savedir was not specified and the cache was the default cache directory. There could be a weird error if they depend on those files being up-to-date and we updated a recipe and the updates aren't reflected in the stale cache, but honestly I would expect that to be rare. If we're concerned about it, we could add a function to check if the cache dir exists and emit a warning.

It's probably fine. A check could make sense if it doesn't have a risk of false positives (which it shouldn't?)

> Overall I'd be happy with this approach, should I go ahead and add it to the template fix PR? Or no, it would probably be better as a separate PR yeah?

I had started to mock up some changes in a `fetch-take-two` branch on my fork. I'm less familiar with the Pretrained API expectations and such. Do you want to continue based on that?

https://github.com/asumagic/speechbrain/tree/fetch-take-two

**Comment by pplantinga** [comment]
Solved by #2711
